File: ios_storyboard/TodoApp/Features/Todo/Data/TodoRepository.swift
import Foundation

@MainActor
final class TodoRepository: TodoRepositoryInterface {
    private let local: TodoLocalDataSource

    init(local: TodoLocalDataSource = TodoLocalDataSource()) {
        self.local = local
    }

    func fetchAll() -> [Todo] { local.fetchAll() }

    func save(_ todo: Todo) { local.save(todo) }

    func delete(id: String) { local.delete(id: id) }

    func toggle(id: String) {
        var todos = local.fetchAll()
        guard let i = todos.firstIndex(where: { $0.id == id }) else { return }
        todos[i].isCompleted.toggle()
        local.save(todos[i])
    }
}

File: android_compose/app/src/main/java/com/example/myapplication/features/todo/TodoRepository.kt
package com.example.myapplication.features.todo

import com.example.myapplication.features.todo.data.local.TodoLocalDataSource
import com.example.myapplication.features.todo.data.remote.TodoRemoteDataSource
import com.example.myapplication.features.todo.data.remote.toEntity
import kotlinx.coroutines.flow.Flow

class TodoRepository(
    private val local: TodoLocalDataSource,
    private val remote: TodoRemoteDataSource = TodoRemoteDataSource(),
) : TodoRepositoryInterface {

    override fun observeAll(): Flow<List<Todo>> = local.observeAll()

    override suspend fun insert(todo: Todo) = local.insert(todo)

    override suspend fun update(todo: Todo) = local.update(todo)

    override suspend fun delete(id: String) = local.delete(id)

    override suspend fun sync() {
        val dtos = remote.fetchAll()
        val existing = local.findAllIds().toSet()
        dtos.filter { "api_${it.id}" !in existing }
            .forEach { local.insert(it.toEntity()) }
    }
}
